pandas/formats/format.py

```diff
@@ -489,30 +489,48 @@ def _to_str_columns(self):
         str_index = self._get_formatted_index(frame)
         str_columns = self._get_formatted_column_labels(frame)
 
-        if self.header:
+        has_aliases = isinstance(self.header, (tuple, list, np.ndarray, Index))
+        if not (has_aliases or self.header):
             stringified = []
             for i, c in enumerate(frame):
-                cheader = str_columns[i]
-                max_colwidth = max(self.col_space or 0, *(self.adj.len(x)
-                                                          for x in cheader))
                 fmt_values = self._format_col(i)
                 fmt_values = _make_fixed_width(fmt_values, self.justify,
-                                               minimum=max_colwidth,
+                                               minimum=(self.col_space or 0),
+                                               adj=self.adj)
+                stringified.append(fmt_values)
+        elif has_aliases:
```

Comment:
It might be worth noting that the only difference between the ``elif has_aliases:`` and ``else:``  conditional branches is the test for raising a ``ValueError`` and ``cheader = [self.header[i]]`` instead of ``cheader = str_columns[i]``.

It would be possible to write instead:
```python
if not (has_aliases or self.header):
    # same code block
    pass
else:
    if has_aliases:
        if len(self.header) != len(self.columns):
            raise ValueError(('Writing %d cols but got %d aliases'
                              % (len(self.columns), len(self.header))))
    stringified = []
    for i, c in enumerate(frame):
        if has_aliases:
            cheader = [self.header[i]]
        else:
            cheader = str_columns[i]
          
        header_colwidth = max(self.col_space or 0,
                              *(self.adj.len(x) for x in cheader))
        fmt_values = self._format_col(i)
        fmt_values = _make_fixed_width(fmt_values, self.justify,
                                       minimum=header_colwidth,
                                       adj=self.adj)

        max_len = max(np.max([self.adj.len(x) for x in fmt_values]),
                      header_colwidth)
        cheader = self.adj.justify(cheader, max_len, mode=self.justify)
        stringified.append(cheader + fmt_values)
```
What do you prefer? Redundancy vs. readability.